### PR TITLE
migrations: Support sane concurrent index creation

### PR DESCRIPTION
**Pull request by efritz**
Pulled from #29831. This PR supports `CREATE INDEX CONCURRENTLY ...` operations without risk of a live-lock/deadlock condition in the presence of concurrently running migrators. This PR also adds code to attempt to repair invalid indexes (which should reduce the number of easily-solvable "dirty database" conditions).

Previously, the migrator will take an advisory lock (if it's confident it needs to run queries) and run all definitions in one shot. Now, the migrator will take an advisory lock and run all definitions that don't require concurrent index creation. Once a concurrent index creation query is encountered, the lock is dropped and the migrator instances will coordinate on the progress of the creation operation (by querying Postgres catalog tables).

There are now two ways that migrators can infer progress from another:

1. If there are pending migrations, a migrator holding a lock is (likely) working on them
2. If there are pending migrations but no migrator holding a lock, there may be an active index creation query
3. If there are pending migrations but neither of those conditions hold, the migration has failed (migrator died)

### CODE REVIEW

File: internal/database/migration/runner/run.go [hunk 1]
```diff
@@ -80,23 +84,42 @@ func (r *Runner) runSchema(ctx context.Context, operation MigrationOperation, sc
...
-) error {
-	up := operation.Type == MigrationOperationTypeTargetedUp
+) (retry bool, _ error) {
+	var (
+		droppedLock bool
+		up          = operation.Type == MigrationOperationTypeTargetedUp
+	)
 
-	callback := func(schemaVersion schemaVersion, _ definitionsByState) error {
+	callback := func(schemaVersion schemaVersion, _ definitionsByState, earlyUnlock unlockFunc) error {
```
Comment (mrnugget) [inline internal/database/migration/runner/run.go:118]: Earl Unlock

screenshot_2022-02-07_15 05 38@2x

**Review by mrnugget — APPROVED** [review]
This is what we talked about on Friday, right?

File: internal/database/migration/runner/run.go [hunk 2]
```diff
@@ -179,3 +354,31 @@ func filterAppliedDefinitions(
 
 	return filtered
 }
+
+// renderIndexStatus returns a slice of interface pairs describing the given index status for use in a
+// call to logger. If the index is currently being created, the progress of the create operation will be
+// summarized.
+func renderIndexStatus(progress storetypes.IndexStatus) (logPairs []interface{}) {
```
Comment (chrismwendt) [inline internal/database/migration/runner/run.go:361]: Found the graphics

File: internal/database/migration/runner/runner.go [hunk 3]
```diff
@@ -229,15 +234,21 @@ func (r *Runner) withLockedSchemaState(
...
-		return err
+	// Detect failed migrations, and determine if we need to wait longer for concurrent migrator
+	// instances to finish their current work.
+	if retry, err := validateSchemaState(ctx, schemaContext, byState); err != nil {
+		return false, err
+	} else if retry {
+		// An index is currently being created. Wait a small time and return to the caller to
+		// be re-invoked. We don't want to take any action here while the other proceses is
+		// working.
+		return true, wait(ctx, indexPollInterval)
```
Comment (chrismwendt) [inline internal/database/migration/runner/runner.go:249]: Nit: this `wait()` is pretty far from the `for` loops where it's relevant, so I'd consider moving it inside the loops. When I saw `for { applyMigrations() }` I wondered if that would get called a lot and accidentally put load on Postgres or spike CPU.

File: internal/database/migration/runner/testdata/concurrent-index/10001/up.sql [hunk 4]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE orders (
+    id int NOT NULL,
+    item_id int NOT NULL,
+    quantity int NOT NULL,
+    order_date TIMESTAMP WITH TIME ZONE,
+    charged boolean NOT NULL DEFAULT false,
+    fulfilled boolean NOT NULL DEFAULT false
+);
```
Comment (chrismwendt) [inline internal/database/migration/runner/testdata/concurrent-index/10001/up.sql:8]: Missed opportunity for a 🏇 table

File: internal/database/migration/runner/util.go [hunk 5]
```diff
@@ -36,24 +37,64 @@ func groupByState(schemaVersion schemaVersion, definitions []definition.Definiti
...
+		return false, newDirtySchemaError(schemaContext.schema.Name, byState.failed)
 	}
 
 	if len(byState.pending) > 0 {
-		// Explicit failures require administrator intervention
-		return newDirtySchemaError(schemaContext.schema.Name, byState.pending)
+		// We are currently holding the lock, so any migrations that are "pending" are either
+		// dead, or they're concurrent index creation objects. We'll partition this set into
+		// those two groups and determine what to do.
+		if pendingDefinitions, failedDefinitions, err := partitionPendingMigrations(ctx, schemaContext, byState.pending); err != nil {
```
Comment (chrismwendt) [inline internal/database/migration/runner/util.go:57]: I got tripped up by the term "dead" here, not sure what it means. I'd guess "failed".

**Review by chrismwendt — APPROVED** [review]

File: internal/database/migration/runner/util.go (hunk 5 above)
Comment (efritz) [reply internal/database/migration/runner/util.go:57]: It actually means abandoned; migrator died before it finished. Will update docs here.

File: internal/database/migration/runner/runner.go (hunk 3 above)
Comment (efritz) [reply internal/database/migration/runner/runner.go:247]: Updated.

File: internal/database/migration/runner/util.go (hunk 5 above)
Comment (efritz) [reply internal/database/migration/runner/util.go:57]: Updated.

**Pull request merged**